core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala

```diff
@@ -124,11 +125,22 @@ private[spark] class ExternalSorter[K, V, C](
     if (shouldPartition) partitioner.get.getPartition(key) else 0
   }
 
+  private val metaInitialRecords = 256
+  private val kvChunkSize = conf.getInt("spark.shuffle.sort.kvChunkSize", 1 << 22) // 4 MB
+  private val useSerializedPairBuffer =
+    !ordering.isDefined && conf.getBoolean("spark.shuffle.sort.serializeMapOutputs", true) &&
+    ser.isInstanceOf[KryoSerializer] &&
```

Comment:
Is there a more general property of serializers that we could test instead of having something Kryo-specific?  Perhaps serializers could implement a method that returns true if their output has the required properties needed by this new shuffle path; this would allow custom serializers to benefit from this change as well (such as Spark SQL's SQLSerializer / SQLSerializer2).